cpp/src/gandiva/gdv_function_stubs.cc

```diff
@@ -635,30 +638,31 @@ const char* gdv_fn_initcap_utf8(int64_t context, const char* data, int32_t data_
   int32_t out_char_len = 0;
   int32_t out_idx = 0;
   uint32_t char_codepoint;
+
+  // Any character is considered as space, except if it is alphanumeric
   bool last_char_was_space = true;
 
   for (int32_t i = 0; i < data_len; i += char_len) {
     char_len = gdv_fn_utf8_char_length(data[i]);
-    // For single byte characters:
-    // If it is a lowercase ASCII character, set the output to its corresponding uppercase
-    // character; else, set the output to the read character
+    // An optimization for single byte characters:
     if (char_len == 1) {
```

Comment:
what happens if we have invalid bytes?

Reply:
[There is a check](https://github.com/apache/arrow/blob/9a2c010706f1822cac3156a3b594206f07f656e6/cpp/src/gandiva/gdv_function_stubs.cc#L675) to see if the character is valid.

[There are some tests](https://github.com/apache/arrow/blob/9a2c010706f1822cac3156a3b594206f07f656e6/cpp/src/gandiva/gdv_function_stubs_test.cc#L395) using invalid bytes too.

Do you see any corner case about invalid bytes that I am missing to deal with in that function?

Reply:
Ah thanks. Missed that since it was not in this change.
nit: you can just check with if ((unsigned char) data[i] > 0) instead of calling the char_length function. Also can you add marker FORCE_INLINE for all these helper functions.

Reply:
@projjal about the check, I use the `char_length` variable to walk through the char array in the `for-loop`, so I think I can keep the check.

Reply:
@projjal about using the `FORCE_INLINE` macro, a problem happened when I executed the unit tests:
```
error: always_inline function might not be inlinable [-Werror=attributes]
  431 | int32_t gdv_fn_utf8_char_length(char c) {
      |         ^~~~~~~~~~~~~~~~~~~~~~~
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
```

Reply:
For FORCE_INLINE macro it looks like you also need to declare the function as inline along with the always inline attribute. Similar to here https://github.com/apache/arrow/blob/master/cpp/src/parquet/murmur3.cc#L38

Reply:
@projjal it will enter in [this check](https://github.com/apache/arrow/blob/9a2c010706f1822cac3156a3b594206f07f656e6/cpp/src/gandiva/gdv_function_stubs.cc#L675) and the function will send an error to the user

Reply:
@projjal the FORCE_INLINE tip apparently worked, I applied the changes about it

Reply:
calculate char_length for non-ascii case. For ascii just set char_len = 1 for iterating.

Reply:
Suggestion applied